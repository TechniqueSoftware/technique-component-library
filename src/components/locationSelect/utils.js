export const urlClubIdRegex = id => new RegExp(`^.*\/${id}(\/)?(#)?(((\/)?(#){1}){1}([0-9a-zA-Z])+?)*?$`, 'gm') // eslint-disable-line no-useless-escape, max-len
const urlLocationIdRegex = id => new RegExp(`^.*\/${id}\/`, 'gm') // eslint-disable-line no-useless-escape

export const getLocationIdFromUrl = (path, clubId) => {
  try {
    const [urlBeginning] = path.match(urlLocationIdRegex(clubId))
    const locationId = path.substring(urlBeginning.length).replace(/[^0-9]/g, '')
    return locationId ? parseInt(locationId, 10) : null
  } catch (err) {
    return null
  }
}
