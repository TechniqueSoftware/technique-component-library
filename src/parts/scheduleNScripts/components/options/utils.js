export const followUpTimeToObject = followUpTime => {
  const getHour = h => {
    if (h === '00') return 12
    const parsedHour = parseInt(h, 10)
    return parsedHour > 12 ? parsedHour - 12 : parsedHour
  }
  const time = followUpTime.split(':')
  const [oldHour, oldMinute] = time
  const hour = getHour(oldHour)
  const minute = parseInt(oldMinute, 10)
  const meridiem = oldHour >= 12 ? 'PM' : 'AM'
  return { hour, minute, meridiem }
}

export const objectToFollowUpTime = (hour, minute, meridiem) => {
  const parseHour = h => meridiem === 'PM' ? parseInt(h, 10) + 12 : h
  return `${parseHour(hour === 12 ? '00' : hour)}:${minute === 0 ? '00' : minute}:00`
}
