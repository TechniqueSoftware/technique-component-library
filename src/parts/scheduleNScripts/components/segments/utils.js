import { FOLLOWUP_STATUS } from './constants'

export const parseTitle = followUpStatus => {
  switch (followUpStatus) {
    case FOLLOWUP_STATUS.WEB_LEAD:
      return 'Web Lead'
    case FOLLOWUP_STATUS.APPT_BOOKED:
      return 'Appt Booked'
    case FOLLOWUP_STATUS.MEMBER:
      return 'Member'
  }
}

export const parseLocations = segmentLocations => segmentLocations.map(({ locationName, locationId }) => ({
  label: `${locationName}`,
  value: locationId
}))

export const checkPermission = userPermission => userPermission.some(
  element => element.permissionName === 'CopySegmentsAction' && element.permissionEffect === 'ALLOW')
