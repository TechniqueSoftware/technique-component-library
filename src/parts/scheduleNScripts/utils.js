import { PERMISSION_KEYS, STATUSES_WITH_SEGMENTS } from './constants'

export const checkIfSegmentStatus = (type, id) => {
  const hasSegmentation = !!STATUSES_WITH_SEGMENTS.find(s => s.type === parseInt(type, 10) && s.id === parseInt(id, 10))
  return hasSegmentation
}

const setSchedulesDayNum = schedules => schedules.reduce(
  (acc, curr) => {
    const dayNum = curr.daysTillNext + acc.dayNum
    return { schedules: acc.schedules.concat({ ...curr, dayNum }), dayNum }
  },
  { dayNum: 0, schedules: [] }
).schedules

export const getStatusSegments = (type, id, segments) => {
  if (!checkIfSegmentStatus(type, id)) return []
  const currentStatus = segments.find(
    ({ followUpType, followUpStatus }) => followUpType === parseInt(type, 10) && followUpStatus === parseInt(id, 10)
  )
  if (!currentStatus) return []
  const { followUpSegments = [], followUpSchedules = [], followUpSegmentScripts = [] } = currentStatus
  return followUpSegments.map(segment => {
    const schedulesWithScripts = followUpSchedules.reduce((acc, schedule) => {
      if (segment.segmentId !== schedule.segmentId) return acc
      return acc.concat({
        ...schedule,
        scripts: followUpSegmentScripts.filter(
          script => script.followUpSeqNum === schedule.followUpSeqNum && script.segmentId === schedule.segmentId
        )
      })
    }, [])
    return { ...segment, followUpSchedules: setSchedulesDayNum(schedulesWithScripts) }
  })
}

export const sortSegments = segments => [...segments].sort((a, b) => a.name.localeCompare(b.name))

// Formatting club-os data into something usable
export const parseSnSData = (data, followUpSchedules, followUpScripts, segmentsData, isClub) => {
  const schedulesPermissions = followUpSchedules.permissions
  const scriptsPermissions = followUpScripts.permissions
  const schedules = isClub ? followUpSchedules : followUpSchedules.schedules
  const scripts = isClub ? followUpScripts : followUpScripts.scripts

  const group = (acc, curr) => {
    const { followUpStatus, followUpType } = curr
    const key = `${followUpStatus}-${followUpType}`
    if (acc[key]) acc[key].push(curr)
    else acc[key] = [curr]
    return acc
  }

  const sch = schedules.reduce(group, {})
  const scr = scripts.reduce(group, {})

  Object.keys(sch).forEach(key => { sch[key] = setSchedulesDayNum(sch[key]) })

  Object.keys(scr).forEach(key => {
    const scripts = scr[key]
    scripts.forEach(script => {
      const schedule = sch[key]?.find(el => el.followUpSeqNum === script.followUpSeqNum)
      if (!schedule) return
      if (schedule.scripts) schedule.scripts.push(script)
      else schedule.scripts = [script]
    })
  })

  const followups = data.map(el => {
    const { type, statuses, id } = el
    const canEditSchedules = isClub ? undefined : schedulesPermissions[PERMISSION_KEYS.SCHEDULES[type]]
    const canEditScripts = isClub ? undefined : scriptsPermissions[PERMISSION_KEYS.SCRIPTS[type]]
    const newStatuses = statuses.map(st => {
      const key = `${st.id}-${type}`
      const unsortedSegments = getStatusSegments(type, st.id, segmentsData)
      const segments = sortSegments(unsortedSegments)
      return { ...st, followUpTab: id, schedules: sch[key] || [], segments }
    })
    return { ...el, statuses: newStatuses, canEditSchedules, canEditScripts }
  })

  return {
    followups,
    canEditSchedules: followups[0].canEditSchedules,
    canEditScripts: followups[0].canEditScripts
  }
}

export const getSchedulesByDay = (schedules, dayNum, order) => {
  const { earlierSch, sameDaySch, laterSch } = schedules.reduce(
    (acc, sch) => {
      if (sch.dayNum < dayNum) acc.earlierSch = acc.earlierSch.concat(sch)
      if (sch.dayNum === dayNum) acc.sameDaySch = acc.sameDaySch.concat(sch)
      if (sch.dayNum > dayNum) acc.laterSch = acc.laterSch.concat(sch)
      return acc
    },
    { earlierSch: [], sameDaySch: [], laterSch: [] }
  )
  return {
    earlierSch: order === 'first' ? earlierSch : earlierSch.concat(sameDaySch),
    laterSch: order === 'first' ? sameDaySch.concat(laterSch) : laterSch
  }
}

export const getCurrentSchedules = (segmentId, status) => {
  if (!segmentId) return status.schedules
  return status.segments.find(segment => segment.segmentId === segmentId).followUpSchedules
}

export const getStatus = (followups, type, status) => {
  const fs = followups.filter(f => f.type === type)
  if (!fs.length) return null

  let st = null
  for (let i = 0; i < fs.length; i++) {
    st = fs[i].statuses.find(s => s.id === `${status}`)
    if (st) break
  }
  return st
}

export const setFollowUpStatusValue = (followUps, status, value) => {
  const newFollowups = followUps.map(followUp => {
    if (followUp.id !== status.followUpTab) return followUp
    const newStatuses = followUp.statuses.map(st => {
      if (st.id !== status.id) return st
      return { ...st, ...value }
    })
    return { ...followUp, statuses: newStatuses }
  })
  return newFollowups
}

export const setFollowUpScripts = (followUps, { followUpType, followUpStatus }, scriptsData, segmentId) => {
  const status = getStatus(followUps, followUpType, followUpStatus)
  const currentSchedules = getCurrentSchedules(segmentId, status)
  const schedules = currentSchedules.map(schedule => {
    if (schedule.followUpSeqNum !== scriptsData.followUpSeqNum) return schedule
    return scriptsData
  })
  const statusValueToUpdate = !segmentId
    ? { schedules }
    : {
      segments: status.segments.map(
        segment => (segment.segmentId !== segmentId ? segment : { ...segment, followUpSchedules: schedules }) // eslint-disable-line max-len
      )
    }
  return setFollowUpStatusValue(followUps, status, statusValueToUpdate)
}

export const updateTouchpointInFollowups = (followUps, tp, data) => {
  const { followUpType, followUpStatus, followUpSeqNum } = tp
  const status = getStatus(followUps, followUpType, followUpStatus)
  const newSchedules = status.schedules.map(sch => {
    if (sch.followUpSeqNum !== followUpSeqNum) return sch
    const newSch = { ...sch, ...data }
    return newSch
  })
  return setFollowUpStatusValue(followUps, status, { schedules: newSchedules })
}
