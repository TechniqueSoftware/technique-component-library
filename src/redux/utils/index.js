export { default as Provider } from './provider'
export { default as TestProvider } from './testProvider'
export { default as recordSaga } from './recordSaga'
