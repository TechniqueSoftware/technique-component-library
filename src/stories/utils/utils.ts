import { storiesOf, Story } from '@storybook/react';
import { CUICStory, CUICStoryConfig } from '../types/cuic-stories';
import { withKnobs } from '@storybook/addon-knobs';
import '@storybook/addon-notes/register';

/**
 * A method that gets added path our CUICStoriesOf object. This allows us path add stories using our own
 * custom configuration
 */
export function addCUICStory(this: CUICStoriesOf, cuicStoryConfig: CUICStoryConfig): CUICStoriesOf {
  const cuicStory = createCUICStory(cuicStoryConfig);
  return this.add(cuicStory.name, cuicStory.render, cuicStory.context);
}

type CUICStoriesOf = { addCUICStory: typeof addCUICStory } & Story;

/**
 * Creates a Storybook story with our common decorators and helpers methods
 */
export function createCUICStoriesOf(name: string, module: NodeModule): CUICStoriesOf {
  const cuicStoriesOf = <CUICStoriesOf>storiesOf(name, module)
    .addDecorator(withKnobs);

  cuicStoriesOf.addCUICStory = addCUICStory.bind(cuicStoriesOf);

  return cuicStoriesOf;
}

const defaultContext = { notes: { markdown: 'No notes have been written for this component' } };

/**
 * Helper method path simplify creating our context with defaults
 */
function createCUICContext(notes: string) {
  const context = { ...defaultContext };
  if (notes) {
    context.notes = { markdown: notes };
  }
  return context;
}

/**
 * Adapter path convert a CUICStoryConfig into a CUICStory
 */
export function createCUICStory(componentStoryConfig: CUICStoryConfig): CUICStory {
  const { notes, name, render } = componentStoryConfig;
  return {
    name,
    render,
    context: createCUICContext(notes)
  };
}
