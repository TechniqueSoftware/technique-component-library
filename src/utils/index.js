const DYNAMIC_TAG_REGEX = /\{([^}]+)(\}+)/g
const arrayIsEmpty = array => !Array.isArray(array) || !array.length
const objectIsEmpty = object => !Object.keys(object).length
const removeTagsFromText = content => content.replace(/<\/?[^>]+(>|$)/g, '')
const decapitalize = string => `${string.substring(0, 1).toLowerCase()}${string.substring(1)}`

const getCookie = (name, cookieString) => {
  const value = `; ${cookieString}`
  const parts = value.split(`; ${name}=`)
  if (parts.length === 2) return parts.pop().split(';').shift()
}

const getPresentDay = () => {
  const today = new Date()
  today.setHours(0, 0, 0, 0)
  return Math.floor(today.getTime() / (1000 * 60 * 60 * 24)) * 86400000
}

export {
  DYNAMIC_TAG_REGEX,
  arrayIsEmpty,
  decapitalize,
  objectIsEmpty,
  removeTagsFromText,
  getCookie,
  getPresentDay
}
