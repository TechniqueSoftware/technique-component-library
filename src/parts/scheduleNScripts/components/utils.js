import { STATUSES_WITH_SEGMENTS } from '@parts/scheduleNScripts/constants'
// SEGMENTS
export const checkIfAttributeIsInUse = (segments, selectedAttributes, valueToCheck, segmentId) => {
  if (valueToCheck === '') return false
  const listToCheck = segments.reduce(
    (acc, segment) => {
      if (segmentId && segment.segmentId === segmentId) return acc
      return acc.concat(segment.segmentAttributes)
    },
    selectedAttributes.map(({ value }) => ({ attributeValue: value }))
  )
  return !!listToCheck.find(({ attributeValue }) => attributeValue === valueToCheck)
}

export const getSegmentStatusProp = ({ followUpType, followUpStatus }, keyword) => {
  const segmentTp = STATUSES_WITH_SEGMENTS.find(
    s => s.type === parseInt(followUpType, 10) && s.id === parseInt(followUpStatus, 10)
  )
  return segmentTp?.[keyword]
}

export const parseAttributes = (attributes, type) => attributes.reduce((acc, attr) => {
  if (!attr.value) return acc
  return acc.concat({ attributeValue: attr.value, attributeType: type, attributeId: attr.id })
}, [])

export const unparseAttributes = attributes => attributes.map(({ attributeValue, attributeId }) => ({
  name: attributeValue,
  value: attributeValue,
  id: attributeId
}))
