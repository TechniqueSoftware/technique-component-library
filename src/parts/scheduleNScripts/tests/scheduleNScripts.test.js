import React from 'react'
import * as redux from 'react-redux'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { TestProvider as Provider } from '@redux/utils'

import SnS from '..'

const followups = [
  {
    id: '0',
    name: 'DragonBall',
    statuses: [
      {
        id: '1',
        name: 'Goku',
        displayName: 'Goku',
        schedules: [
          {
            autoEnabled: false,
            clubId: 659,
            daysTillNext: 2,
            followUpSeqNum: 1,
            dayNum: 2,
            followUpStatus: 11,
            followUpType: 1,
            canEditSchedules: false,
            scripts: [
              {
                actionType: 1,
                clubId: 659,
                followUpSeqNum: 1,
                followUpStatus: 11,
                followUpType: 1,
                script: 'Hi'
              }
            ]
          }
        ]
      },
      {
        id: '2',
        name: 'Vegeta'
      },
      {
        id: '3',
        name: 'Gohan'
      },
      {
        id: '4',
        name: 'Piccolo'
      }
    ]
  },
  {
    id: '1',
    name: 'Naruto',
    statuses: [
      {
        id: '1',
        name: 'Sasuke',
        schedules: [
          {
            autoEnabled: false,
            clubId: 659,
            daysTillNext: 2,
            followUpSeqNum: 1,
            dayNum: 2,
            followUpStatus: 11,
            followUpType: 1,
            canEditSchedules: true,
            scripts: [
              {
                actionType: 1,
                clubId: 659,
                followUpSeqNum: 1,
                followUpStatus: 11,
                followUpType: 1,
                script: 'Hi'
              }
            ]
          }
        ]
      },
      {
        id: '2',
        name: 'Kakashi'
      },
      {
        id: '3',
        name: 'Itachi'
      },
      {
        id: '4',
        name: 'Sakura'
      }
    ]
  },
  {
    id: '2',
    name: 'One Piece',
    statuses: [
      {
        id: '1',
        name: 'Monkey D. Luffy'
      },
      {
        id: '2',
        name: 'Nami'
      },
      {
        id: '3',
        name: 'Usopp'
      },
      {
        id: '4',
        name: 'Chopper'
      },
      {
        id: '5',
        name: 'Franky'
      }
    ]
  }
]

const scheduleNScripts = {
  loading: false,
  error: '',
  followups,
  canEditSchedules: true,
  canEditScripts: true
}

describe('<ScheduleAndScripts />', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(
      <Provider>
        <SnS />
      </Provider>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  describe('Editable', () => {
    it('has scheduels editables', () => {
      jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb({ scheduleNScripts }))
      render(<Provider><SnS /></Provider>)
      const icon = screen.getByText('Schedules are editable').nextElementSibling
      expect(icon).toHaveClass('text-success')
      const icons = screen.queryAllByLabelText('BsCheckLg')
      expect(icons).toContainEqual(icon)
    })

    it('has scripts editables', () => {
      jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb({ scheduleNScripts }))
      render(<Provider><SnS /></Provider>)
      const icon = screen.getByText('Scripts are editable').nextElementSibling
      expect(icon).toHaveClass('text-success')

      const icons = screen.queryAllByLabelText('BsCheckLg')
      expect(icons).toContainEqual(icon)
    })

    it('does not have schedules editable', () => {
      const state = {
        scheduleNScripts: {
          ...scheduleNScripts,
          canEditSchedules: false
        }
      }

      jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb(state))
      render(<Provider><SnS /></Provider>)
      const icon = screen.getByText('Schedules are editable').nextElementSibling
      expect(icon).toHaveClass('text-danger')

      const icons = screen.queryAllByLabelText('BsXLg')
      expect(icons).toContainEqual(icon)
    })

    it('does not have scripts editables', () => {
      const state = {
        scheduleNScripts: {
          ...scheduleNScripts,
          canEditScripts: false
        }
      }

      jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb(state))
      render(<Provider><SnS /></Provider>)
      const icon = screen.getByText('Scripts are editable').nextElementSibling
      expect(icon).toHaveClass('text-danger')

      const icons = screen.queryAllByLabelText('BsXLg')
      expect(icons).toContainEqual(icon)
    })
  })

  it('renders all tabs', () => {
    render(<Provider><SnS /></Provider>)

    followups.forEach(({ name }) => {
      expect(screen.getByText(name)).toBeInTheDocument()
    })
  })

  it('renders all tabs', () => {
    render(<Provider><SnS /></Provider>)

    followups.forEach(({ name }) => {
      expect(screen.getByText(name)).toBeInTheDocument()
    })
    expect(screen.getByText('Schedules are editable')).toBeInTheDocument()
    userEvent.click(screen.getByText('Naruto'))
    expect(screen.getByText('Schedules are editable').nextElementSibling).toHaveClass('text-success')
  })

  it('renders error', () => {
    const sns = {
      ...scheduleNScripts,
      error: 'Error here'
    }
    jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb({ scheduleNScripts: sns }))
    render(<Provider><SnS /></Provider>)

    expect(screen.getByText('Error here')).toBeInTheDocument()
  })
})
