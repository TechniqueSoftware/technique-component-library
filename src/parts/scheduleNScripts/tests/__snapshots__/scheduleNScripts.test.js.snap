// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScheduleAndScripts /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <div
      class="d-flex justify-content-between align-items-center"
    >
      <div>
        <select
          aria-label="Select "
          class="form-select"
          id="locationsSelect"
          role="combobox"
          type="select"
        >
          <optgroup
            label="Locations"
          />
        </select>
      </div>
      <div
        class="d-flex align-items-center gap-5"
      >
        <div
          class="d-flex align-items-center gap-3"
        >
          <p
            class="fw-bold"
          >
            Schedules are editable
          </p>
          <div
            aria-label="BsXLg"
            class="
    d-flex
    justify-content-center
    align-items-center
    text-danger
    
  "
            role="img"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1.293 1.293a1 1 0 0 1 1.414 0L8 6.586l5.293-5.293a1 1 0 1 1 1.414 1.414L9.414 8l5.293 5.293a1 1 0 0 1-1.414 1.414L8 9.414l-5.293 5.293a1 1 0 0 1-1.414-1.414L6.586 8 1.293 2.707a1 1 0 0 1 0-1.414z"
              />
            </svg>
          </div>
        </div>
        <div
          class="d-flex align-items-center gap-3"
        >
          <p
            class="fw-bold"
          >
            Scripts are editable
          </p>
          <div
            aria-label="BsXLg"
            class="
    d-flex
    justify-content-center
    align-items-center
    text-danger
    
  "
            role="img"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1.293 1.293a1 1 0 0 1 1.414 0L8 6.586l5.293-5.293a1 1 0 1 1 1.414 1.414L9.414 8l5.293 5.293a1 1 0 0 1-1.414 1.414L8 9.414l-5.293 5.293a1 1 0 0 1-1.414-1.414L6.586 8 1.293 2.707a1 1 0 0 1 0-1.414z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <hr />
    <div
      class="shadow-sm border-0 card"
    >
      <div
        class="card-body"
      >
        <p
          class="card-text"
        >
          View and edit follow-up schedules and scripts on this page. When selecting a location from the dropdown menu, the access indicators to the right indicate whether scripts and schedules can be edited.
        </p>
        <br />
        <p
          class="card-text"
        >
          Select a tab to view the follow-up statuses for each type. Each box within a follow-up status represents a configured touchpoint scheduled immediately or a number of days after a user enters the status. The email, text, and call icons within each box show what script types are configured, and a clock icon indicates that a follow-up is automatic. Each schedule box has a menu button to manage scripts and schedules for users with permission. Changing a schedule affects newly scheduled follow-ups only.
        </p>
      </div>
    </div>
    <br />
    <div
      class="d-flex justify-content-center"
    >
      <div
        class="spinner-border"
        role="status"
      />
    </div>
  </div>
</DocumentFragment>
`;
